contentful/apps
# Integ 287 slack app does not emit messages in environments other than the configured one

## Purpose

The purpose of this PR is to adjust how the `AuthToken`, held within DynamoDB, is being queried/referenced. Originally it was being referenced with a unique id that was compiled of various attributes, one of them being the `environmentId` that is associated with the current app installation environment. This reference was causing particular issues for environments that are aliases - the configuration page would throw an error due to a lack of credentials accessed via DynamoDB.

## Approach

Given the problem described above, one solution is to instead have an environment agnostic reference to the `AuthToken`, so that all environments within the space can properly reference the appropriate credentials. 

Therefore I implemented the use of an `installationUuid` attribute that is now being used as the reference. This `installationUuid` is added to the DynamoDB table as well as the app installation parameters for persistence. 

Workflow:
1. User installs the Slack app - when connecting to slack, a window opens that takes them through the auth flow. This renders the `useConnect` component, which creates the initial `installationUuid` and eagerly POSTs to the Lamdba, which then creates the appropriate DynamoDb row of data. (this follows the same workflow as before, just with an added `installationUuid` column). 
2. Once the user saves the installation configuration, the `installationUuid` is saved as an app installation parameter. Therefore the `installationUuid` is persisted in the params as well as the DynamoDB. 
3. Once this `installationUuid` exists as a parameter, the backend can pull that information using the CMA, and can then use it to grab the associated `AuthToken` within DynamoDB. 

### Know limitations to this approach: 
1. The user must save the config page to persist the `installationUuid`. If they do not, the auth credentials saved within DynamoDB cannot be accessed, and they will instead have to reinitiate the auth flow (which is fairly easy for the user). The main drawback here is just adding new rows to DynamoDB each time this occurs, so a proposed solution moving forward would be to remind the user to save the page on first install if they try and navigate away or to adjust the flow more significantly to not eagerly create the data in DynamoDB until the config has been saved. (follow-up work, most likely - ticket to come)
2. There is a bug within the CMA/appInstallation workflow that does not correctly provide installation parameters when environments are aliased multiple times. This causes an error to be thrown with the Slack app installation process, but this is very edge case and the plan is to eventually escalate and tackle on DX team with App Framework adjustments. 

## Testing steps

Pull down this code in conjunction with [this](https://github.com/contentful/apps/pull/3634) PR. Within the Slack (development) space created as part of our testing organization, confirm the proper installation and use of the Slack app within environments, that are or are not aliases. In order to test the event triggering functionality (sending slack notifications when publishing content for example), refer to the internal Slack app wiki (would link here if I could - can I?) - ping me with questions!
